File: apps/android/src/main/kotlin/space/bitos/app/ui/create/meme/MemeProjectStore.kt
package space.bitos.app.ui.create.meme

import java.io.File

/**
 * On-disk continuation store for the Quick MEM editor (plan MST-018,
 * EDT-001/002): one dir per slot under `<root>/slots/<slotId>/` holding
 * `slot.json` (shared [space.bitos.core.studio.MemeSlotCodec] wire),
 * copied-in assets (CAP-005: picker URIs are transient — bytes are
 * copied, never referenced) and a `poster.jpg` (~256 px, ≤192 KB). The
 * hub index lives at `<root>/index.json` (shared codec + LRU rules).
 *
 * Platform edges ([BitmapIo], [AssetOpener]) are injected so the
 * round-trip is JVM-testable against a temp dir; production uses
 * [AndroidBitmapIo] + a resolver-backed opener, rooted at
 * `context.filesDir/studio`. Every read is lenient — a corrupt store
 * degrades to "no slots" and never blocks creating.
 */
class MemeProjectStore(
    private val root: File,
    private val bitmaps: BitmapIo = AndroidBitmapIo,
) {

    /** One session asset to persist: project id + opaque source key (uri). */
    data class AssetRef(val id: String, val sourceKey: String)

    /** Opens an asset's bytes by source key; null = unreadable (skipped). */
    fun interface AssetOpener {
        fun open(sourceKey: String): ByteArray?
    }

    /** Bitmap edges the store needs (injected for JVM tests). */
    interface BitmapIo {
        fun aspectOf(bytes: ByteArray): Float?
        fun encodePosterJpeg(bytes: ByteArray, longEdge: Int): ByteArray?
    }

    data class SavedSlot(
        val document: space.bitos.core.studio.MemeSlotDocument,
        /** Asset files keyed by project asset id (resume loads these). */
        val assetFiles: Map<String, File>,
        val posterFile: File?,
    )

    init {
        runCatching { root.mkdirs() }
    }

    // ── Save ─────────────────────────────────────────────────────────────

    /**
     * Persists the slot: copies session assets into the slot dir
     * (idempotent), writes the wire, regenerates the poster, updates the
     * index and LRU-evicts overflow (evicted dirs are deleted here).
     */
    fun save(
        slotId: String,
        projectWire: String,
        assets: List<AssetRef>,
        opener: AssetOpener,
        nowMs: Long,
    ): List<String> {
        val document = space.bitos.core.studio.MemeProjectContract.decode(projectWire)
            ?: return emptyList()
        val slotDir = File(File(root, SLOTS_DIR), slotId).apply { mkdirs() }

        val assetRefs = assets.mapNotNull { ref ->
            val target = File(slotDir, "asset-${ref.id}.img")
            if (!target.exists() || target.length() == 0L) {
                val bytes = opener.open(ref.sourceKey) ?: return@mapNotNull null
                runCatching { target.writeBytes(bytes) }.getOrDefault(Unit)
                if (!target.exists() || target.length() == 0L) return@mapNotNull null
            }
            val bytes = runCatching { target.readBytes() }.getOrNull() ?: return@mapNotNull null
            space.bitos.core.studio.MemeSlotAsset(
                id = ref.id,
                fileName = target.name,
                aspect = bitmaps.aspectOf(bytes) ?: 1f,
            )
        }

        val finalDocument = space.bitos.core.studio.MemeSlotDocument(
            slotId = slotId,
            project = document,
            assets = assetRefs,
            updatedAtMs = nowMs,
        )
        runCatching {
            File(slotDir, SLOT_FILE).writeText(
                space.bitos.core.studio.MemeSlotCodec.encode(finalDocument),
            )
        }

        assetRefs.firstOrNull()?.let { first ->
            File(slotDir, first.fileName).takeIf { it.exists() }?.let { assetFile ->
                runCatching { assetFile.readBytes() }.getOrNull()?.let { bytes ->
                    bitmaps.encodePosterJpeg(bytes, POSTER_LONG_EDGE)?.let { jpeg ->
                        runCatching { File(slotDir, POSTER_FILE).writeBytes(jpeg) }
                    }
                }
            }
        }

        val upsert = space.bitos.core.studio.MemeSlotRules.upsert(
            entries = readIndex(),
            slotId = slotId,
            updatedAtMs = nowMs,
            posterName = POSTER_FILE,
            label = space.bitos.core.studio.MemeSlotRules.labelFor(document),
        )
        writeIndex(upsert.entries)
        upsert.evicted.forEach { evictDir(it) }
        return upsert.evicted
    }

    // ── Read ─────────────────────────────────────────────────────────────

    fun listSlots(): List<space.bitos.core.studio.MemeSlotEntry> = readIndex()

    /** Loads one slot for resume; null when its files are missing/corrupt. */
    fun loadSlot(slotId: String): SavedSlot? {
        val slotDir = File(File(root, SLOTS_DIR), slotId)
        val wire = runCatching { File(slotDir, SLOT_FILE).readText() }.getOrNull() ?: return null
        val document = space.bitos.core.studio.MemeSlotCodec.decode(wire) ?: return null
        val files = document.assets.associate { asset ->
            asset.id to File(slotDir, asset.fileName)
        }
        val poster = File(slotDir, POSTER_FILE).takeIf { it.exists() }
        return SavedSlot(document, files, poster)
    }

    fun posterFile(slotId: String): File? =
        File(File(File(root, SLOTS_DIR), slotId), POSTER_FILE).takeIf { it.exists() }

    // ── Delete ───────────────────────────────────────────────────────────

    /** Removes a slot dir and drops it from the index (explicit deletes only). */
    fun deleteSlot(slotId: String) {
        evictDir(slotId)
        writeIndex(readIndex().filter { it.slotId != slotId })
    }

    private fun evictDir(slotId: String) {
        runCatching { File(File(root, SLOTS_DIR), slotId).deleteRecursively() }
    }

    private fun readIndex(): List<space.bitos.core.studio.MemeSlotEntry> =
        runCatching {
            val indexFile = File(root, INDEX_FILE)
            if (!indexFile.exists()) return emptyList()
            space.bitos.core.studio.MemeSlotCodec.decodeIndex(indexFile.readText())
        }.getOrDefault(emptyList())

    private fun writeIndex(entries: List<space.bitos.core.studio.MemeSlotEntry>) {
        runCatching {
            File(root, INDEX_FILE).writeText(
                space.bitos.core.studio.MemeSlotCodec.encodeIndex(entries),
            )
        }
    }

    companion object {
        const val SLOTS_DIR = "slots"
        const val INDEX_FILE = "index.json"
        const val SLOT_FILE = "slot.json"
        const val POSTER_FILE = "poster.jpg"
        const val POSTER_LONG_EDGE = 256
    }
}

/** Production bitmap edges: BitmapFactory decode + scaled JPEG poster. */
object AndroidBitmapIo : MemeProjectStore.BitmapIo {
    override fun aspectOf(bytes: ByteArray): Float? = runCatching {
        val options = android.graphics.BitmapFactory.Options().apply { inJustDecodeBounds = true }
        android.graphics.BitmapFactory.decodeByteArray(bytes, 0, bytes.size, options)
        if (options.outWidth > 0 && options.outHeight > 0) {
            options.outWidth.toFloat() / options.outHeight
        } else {
            null
        }
    }.getOrNull()

    override fun encodePosterJpeg(bytes: ByteArray, longEdge: Int): ByteArray? = runCatching {
        val source = android.graphics.BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
            ?: return null
        val scale = (longEdge.toFloat() / maxOf(source.width, source.height)).coerceAtMost(1f)
        val poster = if (scale < 1f) {
            android.graphics.Bitmap.createScaledBitmap(
                source,
                (source.width * scale).toInt().coerceAtLeast(1),
                (source.height * scale).toInt().coerceAtLeast(1),
                true,
            )
        } else {
            source
        }
        val out = java.io.ByteArrayOutputStream()
        poster.compress(android.graphics.Bitmap.CompressFormat.JPEG, 80, out)
        if (poster !== source) poster.recycle()
        out.toByteArray()
    }.getOrNull()
}

File: apps/ios/BitOS/Features/Create/MemeProjectStore.swift
import BusinessCore
import CryptoKit
import Foundation
import UIKit

/// On-disk continuation store for the Quick MEM editor (plan MST-018,
/// EDT-001/002): Application Support/studio/slots/<id>/ with slot.json
/// (shared MemeSlotCodec wire), copied-in assets (CAP-005: PhotosPicker
/// items are transient — bytes are copied) and poster.jpg (~256 px, ≤192
/// KB). The index lives at studio/index.json (shared codec + LRU rules).
/// Every read is lenient: a corrupt store degrades to "no slots" and
/// never blocks creating. All IO runs on the caller's background task.
struct MemeProjectStore: Sendable {

    struct SavedSlot {
        let document: MemeSlotDocumentUi
        /// Asset files keyed by project asset id (resume loads these).
        let assetFiles: [String: URL]
        let posterURL: URL?
    }

    /// Display projection of the decoded slot document.
    struct MemeSlotDocumentUi {
        let slotId: String
        let projectJson: String
        let assets: [MemeSlotAssetUi]
        let updatedAtMs: Int64
    }

    struct MemeSlotAssetUi {
        let id: String
        let fileName: String
        let aspect: CGFloat
    }

    struct SlotEntryUi: Identifiable, Equatable {
        let slotId: String
        let updatedAtMs: Int64
        let posterName: String?
        let label: String
        var id: String { slotId }
    }

    private let root: URL

    init(root: URL? = nil) {
        if let root {
            self.root = root
        } else {
            let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            self.root = base.appendingPathComponent("studio", isDirectory: true)
        }
        try? FileManager.default.createDirectory(at: self.root, withIntermediateDirectories: true)
    }

    // MARK: - Paths

    private func slotDir(_ slotId: String) -> URL {
        root.appendingPathComponent("slots", isDirectory: true)
            .appendingPathComponent(slotId, isDirectory: true)
    }

    private var indexURL: URL { root.appendingPathComponent("index.json") }

    // MARK: - Save

    /// Persists the slot: copies session assets in (idempotent), writes
    /// the wire, regenerates the poster, updates the index (LRU evictions
    /// delete their dirs here). Assets are `(id, image)`; the images are
    /// re-encoded as PNG for stable bytes.
    @discardableResult
    func save(
        slotId: String,
        projectJson: String,
        assets: [(id: String, image: UIImage, data: Data?)],
        dataAssets: [(id: String, data: Data, fileName: String)] = [],
        nowMs: Int64
    ) -> [String] {
        let dir = slotDir(slotId)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)

        var assetRows: [[String: Any]] = []
        for asset in assets {
            // MST-053: animated GIF assets persist their ORIGINAL bytes —
            // PNG-encoding would freeze the sticker on its first frame.
            let file = asset.data != nil ? "asset-\(asset.id).gif" : "asset-\(asset.id).png"
            let target = dir.appendingPathComponent(file)
            if !FileManager.default.fileExists(atPath: target.path) {
                if let bytes = asset.data {
                    try? bytes.write(to: target)
                } else if let png = asset.image.pngData() {
                    try? png.write(to: target)
                }
            }
            let size = UIImage(contentsOfFile: target.path)?.size ?? CGSize(width: 1, height: 1)
            let aspect = size.height > 0 ? size.width / size.height : 1
            assetRows.append(["id": asset.id, "file": file, "aspect": aspect])
        }
        // Raw-data assets (video clips): bytes persist verbatim (V1 video).
        for asset in dataAssets {
            let target = dir.appendingPathComponent(asset.fileName)
            if !FileManager.default.fileExists(atPath: target.path) {
                try? asset.data.write(to: target)
            }
            assetRows.append(["id": asset.id, "file": asset.fileName, "aspect": 1.0])
        }

        // Slot wire via the shared codec through the bridge.
        let client = FrameworkBusinessCoreClient()
        let documentJson = buildSlotWire(
            client: client, slotId: slotId, projectJson: projectJson,
            assetRows: assetRows, nowMs: nowMs
        )
        if let documentJson {
            try? documentJson.write(toFile: dir.appendingPathComponent("slot.json").path,
                                    atomically: true, encoding: .utf8)
        }

        // Poster from the first asset (~256 px JPEG).
        if let first = assets.first, let poster = Self.posterJpeg(first.image) {
            try? poster.write(to: dir.appendingPathComponent("poster.jpg"))
        }

        // Index upsert through the shared LRU rules.
        var entries = listSlots()
        let label = labelFor(projectJson: projectJson)
        let bumped = SlotEntryUi(slotId: slotId, updatedAtMs: nowMs, posterName: "poster.jpg", label: label)
        entries.removeAll { $0.slotId == slotId }
        entries.insert(bumped, at: 0)
        let bounded = Array(entries.prefix(MemeSlotsApi.maxSlots))
        let evicted = entries.dropFirst(MemeSlotsApi.maxSlots).map(\.slotId)
        evicted.forEach(deleteSlotFiles)
        try? indexJson(for: bounded).write(toFile: indexURL.path, atomically: true, encoding: .utf8)
        return Array(evicted)
    }

    private func buildSlotWire(
        client: FrameworkBusinessCoreClient,
        slotId: String,
        projectJson: String,
        assetRows: [[String: Any]],
        nowMs: Int64
    ) -> String? {
        // The slot file is the shared MemeSlotCodec shape; compose it as
        // JSON here (project wire embedded verbatim).
        guard let projectData = projectJson.data(using: .utf8),
              let projectObject = try? JSONSerialization.jsonObject(with: projectData) else { return nil }
        let document: [String: Any] = [
            "v": 1,
            "id": slotId,
            "updatedAt": nowMs,
            "assets": assetRows,
            "project": projectObject,
        ]
        guard JSONSerialization.isValidJSONObject(document),
              let data = try? JSONSerialization.data(withJSONObject: document) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    private func labelFor(projectJson: String) -> String {
        // Shared rule: the first overlay text (parsed leniently, no actor
        // isolation needed — the JSON walk is pure).
        guard let data = projectJson.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let rows = root["overlays"] as? [[String: Any]] else { return "Image meme" }
        let first = rows.compactMap { ($0["text"] as? String)?.trimmingCharacters(in: .whitespaces) }
            .first { !$0.isEmpty }
        guard let first else { return "Image meme" }
        let collapsed = first.split(whereSeparator: \.isWhitespace).joined(separator: " ")
        return collapsed.count <= 40 ? String(collapsed) : String(collapsed.prefix(39)) + "…"
    }

    private static func posterJpeg(_ image: UIImage) -> Data? {
        let longEdge = max(image.size.width, image.size.height)
        let scale = min(1, 256 / longEdge)
        let sized = scale < 1
            ? image.resized(scale: scale)
            : image
        return sized.jpegData(compressionQuality: 0.8)
    }

    // MARK: - Read

    func listSlots() -> [SlotEntryUi] {
        guard let json = try? String(contentsOf: indexURL, encoding: .utf8) else { return [] }
        return Self.decodeIndex(json)
    }

    static func decodeIndex(_ json: String) -> [SlotEntryUi] {
        guard let data = json.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              (root["v"] as? NSNumber)?.intValue == 1,
              let rows = root["slots"] as? [[String: Any]] else { return [] }
        let entries: [SlotEntryUi?] = rows.map { row -> SlotEntryUi? in
            guard let id = row["id"] as? String, !id.isEmpty else { return nil }
            var poster: String?
            if let raw = row["poster"] as? String,
               !raw.isEmpty, !raw.hasPrefix("/"), !raw.contains("..") {
                poster = raw
            }
            let labelRaw = (row["label"] as? String) ?? ""
            let label = labelRaw.isEmpty ? "Untitled meme" : labelRaw
            return SlotEntryUi(
                slotId: id,
                updatedAtMs: (row["updatedAt"] as? NSNumber)?.int64Value ?? 0,
                posterName: poster,
                label: label
            )
        }
        return entries.compactMap { $0 }
    }

    private func indexJson(for entries: [SlotEntryUi]) -> String {
        let rows: [[String: Any]] = entries.map { entry in
            var row: [String: Any] = [
                "id": entry.slotId,
                "updatedAt": entry.updatedAtMs,
                "label": entry.label,
            ]
            if let poster = entry.posterName { row["poster"] = poster }
            return row
        }
        guard let data = try? JSONSerialization.data(withJSONObject: ["v": 1, "slots": rows]) else {
            return "{\"v\":1,\"slots\":[]}"
        }
        return String(data: data, encoding: .utf8) ?? "{\"v\":1,\"slots\":[]}"
    }

    func loadSlot(_ slotId: String) -> SavedSlot? {
        let wireURL = slotDir(slotId).appendingPathComponent("slot.json")
        guard let wire = try? String(contentsOf: wireURL, encoding: .utf8),
              let data = wire.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              (root["v"] as? NSNumber)?.intValue == 1,
              let id = root["id"] as? String, !id.isEmpty else { return nil }

        guard let projectObject = root["project"],
              let projectData = try? JSONSerialization.data(withJSONObject: projectObject),
              let projectJson = String(data: projectData, encoding: .utf8) else { return nil }

        var assetFiles: [String: URL] = [:]
        var assetUis: [MemeSlotAssetUi] = []
        for raw in (root["assets"] as? [[String: Any]]) ?? [] {
            guard let assetId = raw["id"] as? String, !assetId.isEmpty,
                  let file = raw["file"] as? String,
                  !file.isEmpty, !file.hasPrefix("/"), !file.contains("..") else { continue }
            let url = slotDir(slotId).appendingPathComponent(file)
            assetFiles[assetId] = url
            assetUis.append(
                MemeSlotAssetUi(
                    id: assetId,
                    fileName: file,
                    aspect: CGFloat((raw["aspect"] as? NSNumber)?.floatValue ?? 1)
                )
            )
        }
        let posterURL = slotDir(slotId).appendingPathComponent("poster.jpg")
        let poster = FileManager.default.fileExists(atPath: posterURL.path) ? posterURL : nil
        return SavedSlot(
            document: MemeSlotDocumentUi(
                slotId: id,
                projectJson: projectJson,
                assets: assetUis,
                updatedAtMs: (root["updatedAt"] as? NSNumber)?.int64Value ?? 0
            ),
            assetFiles: assetFiles,
            posterURL: poster
        )
    }

    func posterURL(_ slotId: String) -> URL? {
        let url = slotDir(slotId).appendingPathComponent("poster.jpg")
        return FileManager.default.fileExists(atPath: url.path) ? url : nil
    }

    // MARK: - Delete

    func deleteSlot(_ slotId: String) {
        deleteSlotFiles(slotId)
        var entries = listSlots()
        entries.removeAll { $0.slotId == slotId }
        try? indexJson(for: entries).write(toFile: indexURL.path, atomically: true, encoding: .utf8)
    }

    private func deleteSlotFiles(_ slotId: String) {
        try? FileManager.default.removeItem(at: slotDir(slotId))
    }
}

/// Shared constants the UI needs (kept in one place next to the store).
enum MemeSlotsApi {
    static let maxSlots = 6
    static let autosaveDebounceMs: UInt64 = 500
}

private extension UIImage {
    func resized(scale: CGFloat) -> UIImage {
        let size = CGSize(width: self.size.width * scale, height: self.size.height * scale)
        let format = UIGraphicsImageRendererFormat()
        format.scale = 1
        return UIGraphicsImageRenderer(size: size, format: format).image { _ in
            draw(in: CGRect(origin: .zero, size: size))
        }
    }
}
